### ISSUE TITLE
FbConfiguration fails with a letter outside the scope of a-z eg. ö

### ISSUE THREAD

**Issue by Rand-Random** [FirebirdSQL/NETProvider#956]
The following code

```C#
static void Main(string[] args)
{
    TestDataBaseName("test.fdb");
    TestDataBaseName("testö.fdb");

    Console.WriteLine("Done");
    Console.ReadKey();

    static void TestDataBaseName(string dataBaseName)
    {
        var conString = GetConnectionString(dataBaseName);
        FbConnection.CreateDatabase(conString, overwrite: true);

        CreateTable(conString);
        InsertData(conString);
        var data = SelectData(conString);
        if (data != 123)
            throw new NotSupportedException();

        ShutDown(conString);
        Online(conString);
    }
    static string GetConnectionString(string dataBaseName)
    {
        var baseDir = AppContext.BaseDirectory;
        var fbConStringBuilder = new FbConnectionStringBuilder()
        {
            DataSource = "localhost",
            Database = Path.Combine(baseDir, dataBaseName),
            UserID = "sysdba",
            Password = "masterkey",
        };

        return fbConStringBuilder.ToString();
    }
    static void CreateTable(string conString)
    {
        using var fbConnection = new FbConnection(conString);
        fbConnection.Open();
        using var fbCommand = fbConnection.CreateCommand();
        fbCommand.CommandText = "create table foo ( id int );";
        _ = fbCommand.ExecuteNonQuery();
    }
    static void InsertData(string conString)
    {
        using var fbConnection = new FbConnection(conString);
        fbConnection.Open();
        using var fbCommand = fbConnection.CreateCommand();
        fbCommand.CommandText = "insert into foo values (123)";
        _ = fbCommand.ExecuteNonQuery();
    }
    static int SelectData(string conString)
    {
        using var fbConnection = new FbConnection(conString);
        fbConnection.Open();
        using var fbCommand = fbConnection.CreateCommand();
        fbCommand.CommandText = "select * from foo";
        return (int)fbCommand.ExecuteScalar();

    }
    static void ShutDown(string conString)
    {
        var fbConfig = new FbConfiguration(conString);
        fbConfig.DatabaseShutdown2(FbShutdownOnlineMode.Single, FbShutdownType.ForceShutdown, 0);
    }
    static void Online(string conString)
    {
        var fbConfig = new FbConfiguration(conString);
        fbConfig.DatabaseOnline2(FbShutdownOnlineMode.Normal);
    }
}
```

will fail with this exception

> Unhandled exception. FirebirdSql.Data.FirebirdClient.FbException (0x80004005): I/O error during "CreateFile (open)" operation for file "C:\USERS\xxx\SOURCE\REPOS\CONSOLEAPP50\CONSOLEAPP50\BIN\RELEASE\NET5.0\TESTö.FDB"
Error while trying to open file
 ---> FirebirdSql.Data.Common.IscException: I/O error during "CreateFile (open)" operation for file "C:\USERS\RFA\SOURCE\REPOS\CONSOLEAPP50\CONSOLEAPP50\BIN\RELEASE\NET5.0\TESTö.FDB"
Error while trying to open file
   at FirebirdSql.Data.Client.Managed.GdsConnection.ProcessResponse(IResponse response)
   at FirebirdSql.Data.Client.Managed.Version10.GdsDatabase.ReadResponse(AsyncWrappingCommonArgs async)
   at FirebirdSql.Data.Client.Managed.Version10.GdsServiceManager.Start(ServiceParameterBuffer spb, AsyncWrappingCommonArgs async)
   at FirebirdSql.Data.Services.FbService.StartTask(ServiceParameterBuffer spb, AsyncWrappingCommonArgs async)
   at FirebirdSql.Data.Services.FbService.StartTask(ServiceParameterBuffer spb, AsyncWrappingCommonArgs async)
   at FirebirdSql.Data.Services.FbConfiguration.DatabaseShutdown2Impl(FbShutdownOnlineMode mode, FbShutdownType type, Int32 seconds, AsyncWrappingCommonArgs async)
   at FirebirdSql.Data.Services.FbConfiguration.DatabaseShutdown2(FbShutdownOnlineMode mode, FbShutdownType type, Int32 seconds)
   at ConsoleApp50.Program.<Main>g__ShutDown|0_5(String conString) in C:\Users\rfa\source\repos\ConsoleApp50\ConsoleApp50\Program.cs:line 73
   at ConsoleApp50.Program.<Main>g__TestDataBaseName|0_0(String dataBaseName) in C:\Users\rfa\source\repos\ConsoleApp50\ConsoleApp50\Program.cs:line 27
   at ConsoleApp50.Program.Main(String[] args) in C:\Users\rfa\source\repos\ConsoleApp50\ConsoleApp50\Program.cs:line 13

As you can see from the code only `FbConfiguration` has an issue with the chracter `ö` a normal `FbConnection` works just fine.
The issue got introduced in version `7.0.0` and is still to be found in `8.0.1`, the issue only happens when connecting to `Firebird Server 2.5.x`.

I am using the following version of Firebird Server
![image](https://user-images.githubusercontent.com/9868626/117692172-0c67da00-b1bd-11eb-9a61-e467c6dc353b.png)

Additionally I have noticed that the error is already present in `6.7.0`, a version prior to the mentioned range of versions I have issues with, if I build the above code with `net5.0` instead of `net47`.

So to sum it up for `Firebird Server 2.5.x`

| FbProvider Version | Target Framework | Issue ( yes / no ) |
| --- | --- | --- |
| 6.7.0 | net5.0 | yes |
| 6.7.0 | net47 | no ( I am here ) |
| 7.0.0 | net47 | yes |
| 7.0.0 | net5.0 | yes |
| 7.10.1 | net47 | yes ( updated to here ) |
| 7.10.1 | net5.0 | yes |
| 8.0.1 | net47 | yes |
| 8.0.1 | net5.0 | yes |

I am coming from `6.7.0` with `net47` and updated to `7.10.1` and still use `net47`, I have only noticed the difference between `net47` and `net5.0` when I wrote this demonstration code.

As a side note
The command `FbConnection.CreateDatabase(conString, overwrite: true);` seem to get stuck when the file already exists, again only seems to be replicatable with `Firebird Server 2.5.x`, so overwriting seem no option.
I therefor added those two lines into my `GetConnectionString` method

```C#
if (File.Exists(fbConStringBuilder.Database))
    File.Delete(fbConStringBuilder.Database);
```

**Comment by InOcobo** [comment]
I use version 7.5 and with the FbBackup command it also fails with the same error.

**Comment by cincuranet** [comment]
IIRC the `isc_spb_utf8_filename` is supported only for FB 3.0+. Before the ANSI was basically used. Compared to `isc_dpb_utf8_filename` that was available in FB 2.5.

**Comment by Rand-Random** [comment]
@InOcobo 
I believe the error is to be found in the base class `FbService`, so every class that derives from this should fail with the same exception.

So all those:
![image](https://user-images.githubusercontent.com/9868626/118115716-9b593a00-b3e9-11eb-96f9-64e75cfd1c56.png)

@cincuranet 
I am curious about what the difference between `.NET47` and `.NET5.0` is going to be.
It kind of hit me off-guard when I prepared the demonstration and the `.NET Provider` version `6.7.0` failed when the demo app targeted `.NET5.0` and I had a "what the fu**" moment and wondered why it is working in the real application.

In the end it was just a lucky guess that changing the target framework of the application could give different results.

Not sure what the difference between `.NET47` and `.NET5.0` could be.
I guess it may have something to do with 
https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.default?view=net-5.0

where the `Encoding.Default` in `.NET47` and `.NET5.0` is different.
But in the end I have no clue, though I will read and try to understand the changes you may commit to fix this issue.

May I ask what the chances are that the error will be fixed in the 7.x branch?
I am currently holding off to update to 8.x because of https://github.com/FirebirdSQL/NETProvider/issues/953

**Comment by cincuranet** [comment]
> @InOcobo
> I believe the error is to be found in the base class `FbService`, so every class that derives from this should fail with the same exception.

Correct.

> I guess it may have something to do with
> https://docs.microsoft.com/en-us/dotnet/api/system.text.encoding.default?view=net-5.0

Yep. Mostly this.

> But in the end I have no clue, though I will read and try to understand the changes you may commit to fix this issue.

Not sure there's an issue here. FB 2.5 does not support `isc_spb_utf8_filename` (I have to double check this) and hence the encoding is expected to be ANSI and there might be a mismatch obviously.

> May I ask what the chances are that the error will be fixed in the 7.x branch?

Virtually zero.

**Comment by cincuranet** [comment]
#711

**Comment by cincuranet** [comment]
OK, I finally understood what's happening in code and what the regression is. Try to use build from [this CI run](https://github.com/FirebirdSQL/NETProvider/actions/runs/842957445) and let me know. Under `net47` it should behave as 6.7.0. On `net5.0` only together with FB3 it will be OK (because of the `Encoding.Default`).

**Comment by Rand-Random** [comment]
@cincuranet 
It is as you said, issue got resolved when I target `net47` but still present on `net5.0`.
Thanks.

Just out of curiosity, not that I need the combination `net5.0` and `FB2.5.x`, is it possible to fix the issue on `net5.0`?
Maybe a naiive idea but shouldn't this be enough to fix it?

```
if (_svc is Client.Managed.Version12.GdsServiceManager)
    SpbFilenameEncoding = Encoding.ANSI;
else
...
```

After writing I noticed there is no `Encoding.ANSI`.
A google session later, I believe this would be necessary to fix `net5.0`
https://stackoverflow.com/questions/62406876

> > May I ask what the chances are that the error will be fixed in the 7.x branch?
> 
> Virtually zero.

To bad.
I will stick with the workaround I have implemented, this is when `FbConfiguration` fails with error code `335544344` I am using the `gfix.exe` tool as a fallback.

Sadly, the error sliped through testing and we shipped a version with this issue, and had to fix it rather urgently, so I did this fallback idea.

Fun note:
The `gfix.exe` tool had the same issue, on shutting down the database, but only with local/embedded - `C:\testö.fbd` - paths, using the TCP/IP scheme - `localhost/3050:C:\testö.fdb` - worked fine.
Though shutting down the database went fine, I guess some cleanup code failed since I got "connection lost to database" which is obvious since the database got shutdown, but I could update the `gfix.exe` tool from `2.5.4` to `2.5.9` which didn't have the issue.
